cpu/stm32_common/periph/i2c_2.c

```diff
@@ -220,25 +220,32 @@ int i2c_read_bytes(i2c_t dev, uint16_t address, void *data, size_t length,
     assert(i2c != NULL);
 
     int ret = 0;
-
     if (!(flags & I2C_NOSTART)) {
         DEBUG("[i2c] read_bytes: Send Slave address and wait for ADDR == 1\n");
         ret = _start(i2c, address, I2C_FLAG_READ, flags);
         if (ret < 0) {
             return ret;
         }
-    }
-
-    if (length == 1) {
-        DEBUG("[i2c] read_bytes: Set ACK = 0\n");
-        i2c->CR1 &= ~(I2C_CR1_ACK);
+        _clear_addr(i2c);
+        if (length == 1 && !(flags & I2C_NOSTOP)) {
+            DEBUG("[i2c] read_bytes: Set ACK = 0\n");
+            i2c->CR1 &= ~(I2C_CR1_ACK);
+        }
+        else {
+            i2c->CR1 |= I2C_CR1_ACK;
+        }
+        _clear_addr(i2c);
```

Comment:
is it necessary to do this twice, i.e., before and after this `if`?

Reply:
Yup, good catch.  I got rid of the first one since it reflects the old code closer.